### PR TITLE
add: stage instances

### PR DESCRIPTION
**Pull request by rigormorrtiss**
Reference: https://github.com/discord/discord-api-docs/pull/2898

### CODE REVIEW

File: src/helpers/misc/delete_stage_instance.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { rest } from "../../rest/rest.ts";
+import { endpoints } from "../../util/constants.ts";
+
+/** Deletes the Stage instance. Requires the user to be a moderator of the Stage channel. */
+export async function deleteStageInstance(channelId: bigint) {
+  // deno-lint-ignore ban-types
+  return await rest.runMethod<{}>(
```
Comment (rigormorrtiss) [inline src/helpers/misc/delete_stage_instance.ts:7]: If the body is not dispatched from the server, the runMethod() returns an empty JSON-parsed object instead.
The type could be changed to `Record<...>` instead, however, it was obscure and ambiguous, hence, `{}`.

File: src/helpers/misc/update_stage_instance.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { rest } from "../../rest/rest.ts";
+import { Errors } from "../../types/discordeno/errors.ts";
+import { StageInstance } from "../../types/misc/stage_instance.ts";
+import { endpoints } from "../../util/constants.ts";
+import { validateLength } from "../../util/validate_length.ts";
+
+/** Updates fields of an existing Stage instance. Requires the user to be a moderator of the Stage channel. */
+export async function updateStageInstance(channelId: bigint, topic: string) {
```
Comment (itohatweb) [inline src/helpers/misc/update_stage_instance.ts:8]: Add proper perm checks

**Review by itohatweb — CHANGES_REQUESTED** [review]
Please move the stage instance helpers to the channel folder.

File: src/helpers/mod.ts [hunk 3]
```diff
@@ -375,6 +383,10 @@ export let helpers = {
   sendMessage,
   unpinMessage,
   // misc
+  createStageInstance,
+  getStageInstance,
+  updateStageInstance,
+  deleteStageInstance,
```
Comment (itohatweb) [inline src/helpers/mod.ts:389]: move this to channels

File: src/helpers/misc/create_stage_instance.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { validateLength } from "../../util/validate_length.ts";
+import { Errors } from "../../types/discordeno/errors.ts";
+import { rest } from "../../rest/rest.ts";
+import { endpoints } from "../../util/constants.ts";
+import { StageInstance } from "../../types/misc/stage_instance.ts";
+
+/** Creates a new Stage instance associated to a Stage channel. Requires the user to be a moderator of the Stage channel. */
+export async function createStageInstance(channelId: string, topic: string) {
```
[the remark this answers is not in the document]
Comment (itohatweb) [reply src/helpers/misc/create_stage_instance.ts:8]: also check if the target channel is a stage channel

File: src/helpers/channels/create_stage_instance.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { validateLength } from "../../util/validate_length.ts";
+import { Errors } from "../../types/discordeno/errors.ts";
+import { rest } from "../../rest/rest.ts";
+import { endpoints } from "../../util/constants.ts";
+import { StageInstance } from "../../types/channels/stage_instance.ts";
```
Comment (itohatweb) [inline src/helpers/channels/create_stage_instance.ts:5]: Use import type to import interfaces
```suggestion
import type { StageInstance } from "../../types/channels/stage_instance.ts";
```

File: src/helpers/channels/get_stage_instance.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { cacheHandlers } from "../../cache.ts";
+import { rest } from "../../rest/rest.ts";
+import { ChannelTypes } from "../../types/channels/channel_types.ts";
+import { StageInstance } from "../../types/channels/stage_instance.ts";
```
Comment (itohatweb) [inline src/helpers/channels/get_stage_instance.ts:4]: Use import type to import interfaces
```suggestion
import type { StageInstance } from "../../types/channels/stage_instance.ts";
```

File: src/helpers/channels/update_stage_instance.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { rest } from "../../rest/rest.ts";
+import { Errors } from "../../types/discordeno/errors.ts";
+import { StageInstance } from "../../types/channels/stage_instance.ts";
```
Comment (itohatweb) [inline src/helpers/channels/update_stage_instance.ts:3]: Use import type to import interfaces
```suggestion
import type { StageInstance } from "../../types/channels/stage_instance.ts";
```

File: src/helpers/channels/create_stage_instance.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import { cacheHandlers } from "../../cache.ts";
+import { ChannelTypes } from "../../types/channels/channel_types.ts";
+import { requireBotChannelPermissions } from "../../util/permissions.ts";
+
+/** Creates a new Stage instance associated to a Stage channel. Requires the user to be a moderator of the Stage channel. */
+export async function createStageInstance(channelId: bigint, topic: string) {
+  const channel = await cacheHandlers.get("channels", channelId);
+  if (!channel) throw new Error(Errors.CHANNEL_NOT_FOUND);
```
Comment (itohatweb) [inline src/helpers/channels/create_stage_instance.ts:13]: Do not throw an error if the channel is not cached instead use `if (channel) // do checks`

File: src/helpers/channels/delete_stage_instance.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import { rest } from "../../rest/rest.ts";
+import { ChannelTypes } from "../../types/channels/channel_types.ts";
+import { Errors } from "../../types/discordeno/errors.ts";
+import { endpoints } from "../../util/constants.ts";
+import { requireBotChannelPermissions } from "../../util/permissions.ts";
+
+/** Deletes the Stage instance. Requires the user to be a moderator of the Stage channel. */
+export async function deleteStageInstance(channelId: bigint) {
+  const channel = await cacheHandlers.get("channels", channelId);
+  if (!channel) throw new Error(Errors.CHANNEL_NOT_FOUND);
```
Comment (itohatweb) [inline src/helpers/channels/delete_stage_instance.ts:11]: Do not throw an error if the channel is not cached instead use `if (channel) // do checks`

File: src/helpers/channels/get_stage_instance.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import { Errors } from "../../types/discordeno/errors.ts";
+import { endpoints } from "../../util/constants.ts";
+
+/** Gets the stage instance associated with the Stage channel, if it exists. */
+export async function getStageInstance(channelId: bigint) {
+  const channel = await cacheHandlers.get("channels", channelId);
+  if (!channel) throw new Error(Errors.CHANNEL_NOT_FOUND);
```
Comment (itohatweb) [inline src/helpers/channels/get_stage_instance.ts:11]: Do not throw an error if the channel is not cached instead use `if (channel) // do checks`

File: src/helpers/channels/update_stage_instance.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import { endpoints } from "../../util/constants.ts";
+import { validateLength } from "../../util/validate_length.ts";
+import { cacheHandlers } from "../../cache.ts";
+import { requireBotChannelPermissions } from "../../util/permissions.ts";
+import { ChannelTypes } from "../../types/channels/channel_types.ts";
+
+/** Updates fields of an existing Stage instance. Requires the user to be a moderator of the Stage channel. */
+export async function updateStageInstance(channelId: bigint, topic: string) {
+  const channel = await cacheHandlers.get("channels", channelId);
+  if (!channel) throw new Error(Errors.CHANNEL_NOT_FOUND);
```
Comment (itohatweb) [inline src/helpers/channels/update_stage_instance.ts:13]: Do not throw an error if the channel is not cached instead use `if (channel) // do checks`

**Review by itohatweb — CHANGES_REQUESTED** [review]

**Review by itohatweb — APPROVED** [review]

**Comment by itohatweb** [comment]
Just talked to skillz, can you add unit tests for these functions?

**Pull request merged**